File: Backend/UtilityBillingSystem.Core/Entities/Bill.cs
using UtilityBillingSystem.Core.Enums;

namespace UtilityBillingSystem.Core.Entities
{
    public class Bill
    {
        public int BillId { get; set; }
        public string BillNumber { get; set; }
        public int ConsumerId { get; set; }
        public int ConnectionId { get; set; }
        public int BillingCycleId { get; set; }
        public int MeterReadingId { get; set; }

        // Amounts
        public decimal Consumption { get; set; }
        public decimal EnergyCharge { get; set; }
        public decimal FixedCharge { get; set; }
        public decimal PreviousOutstanding { get; set; }
        public decimal Subtotal { get; set; }
        public decimal TaxAmount { get; set; }
        public decimal PenaltyAmount { get; set; }
        public decimal TotalAmount { get; set; }
        public decimal PaidAmount { get; set; }
        public decimal OutstandingAmount { get; set; }

        // Dates
        public DateTime BillDate { get; set; }
        public DateTime DueDate { get; set; }
        public DateTime? PaymentDate { get; set; }

        // Status
        public BillStatus Status { get; set; }

        public DateTime CreatedAt { get; set; }
        public int GeneratedBy { get; set; }

        // Navigation
        public Consumer Consumer { get; set; }
        public Connection Connection { get; set; }
        public BillingCycle BillingCycle { get; set; }
        public MeterReading MeterReading { get; set; }
        public User GeneratedByUser { get; set; }
        public ICollection<Payment> Payments { get; set; }
    }
}

File: utility-billing-system/src/app/core/models/bill.ts
export interface Bill {
  billId: number;
  billNumber: string;
  consumerId: number;
  consumerName: string;
  consumerCode: string;
  connectionId: number;
  meterNumber: string;
  utilityTypeName: string;
  cycleName: string;
  consumption: number;
  energyCharge: number;
  fixedCharge: number;
  previousOutstanding: number;
  subtotal: number;
  taxAmount: number;
  penaltyAmount: number;
  totalAmount: number;
  paidAmount: number;
  outstandingAmount: number;
  billDate: string;
  dueDate: string;
  paymentDate: string | null;
  status: 'Generated' | 'Partial' | 'Paid' | 'Overdue';
}

export interface GenerateBillRequest {
  meterReadingId: number;
}